### PR TITLE
Feature/s3 c 1801 policy tag condition keys

### PR DESCRIPTION
**Pull request by dora-korpar**
Necessary changes for evaluation of IAM policies containing tag condition keys. An explanation of the new workflow can be found in this PR: https://github.com/scality/Arsenal/pull/1266
Tests in progress.

### CODE REVIEW

File: lib/api/api.js [hunk 1]
```diff
@@ -100,34 +101,34 @@ const api = {
...
                     }
                 } else {
-                    for (let i = 0; i < authorizationResults.length; i++) {
-                        if (!authorizationResults[i].isAllowed) {
+                    for (let i = 0; i < authResults.length; i++) {
+                        if (!authResults[i].isAllowed) {
                             log.trace('authorization denial from Vault');
                             return callback(errors.AccessDenied);
                         }
                     }
```
Comment (miniscruff) [inline lib/api/api.js]: Looks like a good case for `some`.
```suggestion
                    if (authResults.some(ar => !ar.isAllowed) {
                        log.trace('authorization denial from Vault');
                        return callback(errors.AccessDenied);
                    }
```

File: lib/api/api.js [hunk 2]
```diff
@@ -157,16 +158,32 @@ const api = {
                 // Convert array of post buffers into one string
                 request.post = Buffer.concat(post, postLength).toString();
 
-                if (apiMethod === 'objectCopy' ||
+                console.log(`\n\n--------auth results?????????? ${authorizationResults}`);
```
Comment (miniscruff) [inline lib/api/api.js]: You still need this?

File: lib/api/apiUtils/authorization/tagConditionKeys.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+const { auth, s3middleware } = require('arsenal');
+const metadata = require('../../../metadata/wrapper');
+const { decodeVersionId } = require('../object/versioning');
+
+const { parseTagXml } = s3middleware.tagging;
+
+function makeTagQuery(tags) {
+    const tagsArray = Object.keys(tags);
+    tagsArray.map(t => `${t}=${tags[t]}`);
+    return tagsArray.join(',');
```
Comment (miniscruff) [inline lib/api/apiUtils/authorization/tagConditionKeys.js]: Maaaaybe a bit of a stretch but... 
```suggestion
    return Object.entires(tag)
        .map(([k, v]) => `${k}=${v}`)
        .join(',');
```

File: lib/api/objectPutTagging.js [hunk 4]
```diff
@@ -46,6 +46,8 @@ function objectPutTagging(authInfo, request, log, callback) {
         versionId: reqVersionId,
     };
 
+    const util = require('util');
+    console.log(`\n\n------- REQUEST POST IN PUT OBJ TAGGING::::::::; ${util.inspect(request.post, false, null)}`);
```
Comment (miniscruff) [inline lib/api/objectPutTagging.js]: another log

File: lib/api/apiUtils/authorization/tagConditionKeys.js (hunk 3 above)
Comment (tmacro) [reply lib/api/apiUtils/authorization/tagConditionKeys.js]: alternatively
```
    return Object.entries(tag)
        .map(i => i.join('='))
        .join(',');
```

**Review by tmacro — COMMENTED** [review]
Other than miniscruff's comments LGTM

File: lib/api/apiUtils/authorization/tagConditionKeys.js (hunk 3 above)
Comment (miniscruff) [reply lib/api/apiUtils/authorization/tagConditionKeys.js]: ooh that works too

File: tests/unit/api/apiUtils/tagConditionKeys.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+const requestContexts = [
+    createRequestContext('objectPut', objectPutReq),
+];
+
+describe('tagConditionKeyAuth', () => {
+    it('should return empty if no previous auth results', done => {
+        tagConditionKeyAuth([], objectPutReq, requestContexts, log, err => {
+            assert.ifError(err);
+            done();
+        });
```
Comment (miniscruff) [inline tests/unit/api/apiUtils/tagConditionKeys.js]: I believe you can just pass done into the callback for the same results.

```suggestion
        tagConditionKeyAuth([], objectPutReq, requestContexts, log, done);
```

File: tests/functional/aws-node-sdk/test/object/auth/tagConditionKeys.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        let s3;
+
+        beforeEach(() => {
+            bucketUtil = new BucketUtility('default', sigCfg);
+            s3 = bucketUtil.s3;
+            return s3.createBucketPromise({ Bucket: bucket })
+            .catch(err => {
+                process.stdout.write(`Error creating bucket: ${err}\n`);
+                throw err;
+            });
```
Comment (miniscruff) [inline tests/functional/aws-node-sdk/test/object/auth/tagConditionKeys.js:29]: Not sure how helpful catching an error just to log it is in the context of tests. Wouldn't this just alter the call stack for adding a log statement that we could probably already conclude from the error itself given line numbers and test names?

File: lib/api/api.js [hunk 7]
```diff
@@ -157,16 +162,35 @@ const api = {
...
+                    }
+                    if (tagAuthResults) {
+                        const checkedResults = checkAuthResults(tagAuthResults);
+                        if (checkedResults instanceof Error) {
+                            return callback(checkedResults);
+                        }
+                        returnTagCount = checkedResults;
+                    }
+                    if (apiMethod === 'objectPut' || apiMethod === 'objectPutPart') {
+                        return this[apiMethod](userInfo, request, streamingV4Params,
```
Comment (miniscruff) [inline lib/api/api.js]: I know you didn't make this change in particular, but using `this[apiMethod]` with varying parameters is quite confusing to me. Especially cause I am not even sure where the actual methods are ( I couldn't find them ). I am not sure a solution and I wouldn't block this PR cause of it, but definitely wanted to at least mention it.

File: tests/unit/api/apiUtils/tagConditionKeys.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        tagConditionKeyAuth(authResults, objectPutReq, requestContexts, log, err => {
+            assert.ifError(err);
+            done();
+        });
+    });
+});
+
+describe('updateRequestContexts', () => {
+    before(done => {
+        cleanup();
```
Comment (miniscruff) [inline tests/unit/api/apiUtils/tagConditionKeys.js:64]: cleanup before and after?

**Review by tmacro — APPROVED** [review]

File: tests/unit/api/apiUtils/tagConditionKeys.js (hunk 8 above)
Comment (dora-korpar) [reply tests/unit/api/apiUtils/tagConditionKeys.js:64]: it's standard throughout our test suite 🤷

File: lib/api/api.js (hunk 7 above)
Comment (dora-korpar) [reply lib/api/api.js]: yeah, it's kind of confusing, but the api list is just below the big function in the same file, and each name is the imported method. they're all part of the same object, hence the `this`

File: tests/unit/api/apiUtils/tagConditionKeys.js (hunk 5 above)
Comment (dora-korpar) [reply tests/unit/api/apiUtils/tagConditionKeys.js]: yeah. i kinda prefer having the check written out, though.

File: tests/functional/aws-node-sdk/test/object/auth/tagConditionKeys.js (hunk 6 above)
Comment (dora-korpar) [reply tests/functional/aws-node-sdk/test/object/auth/tagConditionKeys.js:29]: yeah, this is another standard throughout our test suites

File: lib/api/api.js (hunk 7 above)
Comment (miniscruff) [reply lib/api/api.js]: oh ok, I went up not down the file. The review was cut just before the methods.

**Review by miniscruff — APPROVED** [review]

**Comment by dora-korpar** [comment]
@bert-e reset

**Comment by dora-korpar** [comment]
@bert-e wait

**Pull request merged**